src/udp/connection_cookie.rs

```diff
@@ -0,0 +1,214 @@
+
+    for offset in 0..=detail::PERIODS_IN_LIFETIME {
+        let checking_period = detail::get_period_now().saturating_sub(offset as u128);
+
+        let cookie = detail::make_cookie_internal(remote_address, &checking_period);
+
+        if *connection_id == detail::make_connection_id(&cookie) {
+            return Ok(detail::get_time_at_period_start(&checking_period));
+        }
+    }
+    Err(ServerError::InvalidConnectionId)
+}
+
+#[cfg(test)]
+mod tests {
+
+    use std::net::{IpAddr, Ipv4Addr, Ipv6Addr, SocketAddr};
+    use std::time::Duration;
+
+    use self::super::*;
+    use crate::protocol::clock::{StoppedClock, StoppedTime};
+
+    type PeriodLength = Duration;
+
+    fn get_period_length() -> PeriodLength {
+        Duration::from_secs(detail::PERIOD_LENGTH)
+    }
+
+    fn get_last_valid_period_now() -> detail::Period {
+        detail::get_period_now() + detail::PERIODS_IN_LIFETIME as u128
+    }
+
+    fn get_test_socket_addr() -> SocketAddr {
+        SocketAddr::new(IpAddr::V4(Ipv4Addr::new(127, 0, 0, 1)), 8080)
+    }
+
+    fn make_test_cookie(remote_address: Option<&SocketAddr>) -> ConnectionId {
+        make_connection_cookie(remote_address.unwrap_or(&get_test_socket_addr()))
+    }
+
+    #[test]
+    fn it_should_make_a_connection_cookie() {
+        // remote_address: 127.0.0.1:8080, period: 0,
+        // seed: [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
+
+        const ID_STABLE: i64 = 8313098377115836164;
+        const ID_NIGHTLY: i64 = -4342702349221980886;
+
+        let test_cookie = make_test_cookie(None);
+        assert!((test_cookie == ConnectionId(ID_STABLE) || (test_cookie == ConnectionId(ID_NIGHTLY))))
+    }
+
+    #[test]
+    fn it_should_make_different_connection_cookie_with_different_remote_addresses() {
+        let test_remote_address_1 = SocketAddr::new(IpAddr::V6(Ipv6Addr::UNSPECIFIED), 1);
+        let test_remote_address_2 = SocketAddr::new(IpAddr::V6(Ipv6Addr::UNSPECIFIED), 2);
+        let test_remote_address_3 = SocketAddr::new(IpAddr::V6(Ipv6Addr::LOCALHOST), 1);
+
+        assert_ne!(
+            make_test_cookie(Some(&test_remote_address_1)),
+            make_test_cookie(Some(&test_remote_address_2))
+        );
+
+        assert_ne!(
+            make_test_cookie(Some(&test_remote_address_1)),
+            make_test_cookie(Some(&test_remote_address_3))
+        );
+
+        assert_ne!(
+            make_test_cookie(Some(&test_remote_address_2)),
+            make_test_cookie(Some(&test_remote_address_3))
+        )
+    }
+
+    #[test]
+    fn it_should_make_different_cookies_for_the_next_period() {
+        let cookie_now = make_test_cookie(None);
+
+        StoppedClock::local_add(&get_period_length()).unwrap();
+
+        let cookie_next = make_test_cookie(None);
+
+        assert_ne!(cookie_now, cookie_next)
+    }
+
+    #[test]
+    fn it_cookies_should_be_valid_for_this_and_next_periods() {
```

Comment:
@da2ce7 typo: `it_cookies_should_be_valid_for_this_and_next_periods` should be `it_should_be_valid_for_this_and_next_periods`

Reply:
Done :) Thanks :+1: